### PR TITLE
[Module Minifier Plugin] Add support for external modules

### PR DESCRIPTION
**Pull request by dmichon-msft**
Support routing external modules via marking the identifiers as "reserved" to Terser during the initial minification process, then following up after rehydration by converting the `__WEBPACK_EXTERNAL_MODULE_*` identifiers to more compact identifiers.

### CODE REVIEW

File: common/reviews/api/module-minifier-plugin.api.md [hunk 1]
```diff
@@ -46,6 +47,7 @@ export const IDENTIFIER_TRAILING_DIGITS: string;
 
 // @public
 export interface IExtendedModule extends webpack.compilation.Module, webpack.Module {
+    external?: boolean;
```
Comment (iclanton) [inline common/reviews/api/module-minifier-plugin.api.md:50]: Should we call this `isExternal`?

File: webpack/module-minifier-plugin/src/ModuleMinifierPlugin.ts [hunk 2]
```diff
@@ -327,6 +329,11 @@ export class ModuleMinifierPlugin implements webpack.Plugin {
         async (chunks: webpack.compilation.Chunk[]): Promise<void> => {
           // Still need to minify the rendered assets
           for (const chunk of chunks) {
+            const externals: string[] = [];
+            const externalNames: Map<string, string> = new Map();
+            // Skip the first two parameters to not collide with the module function arguments
+            let nextOrdinal: number = 1;
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/ModuleMinifierPlugin.ts:335]: Should we call this `externalIndex` instead?

File: webpack/module-minifier-plugin/src/RehydrateAsset.ts [hunk 3]
```diff
@@ -133,5 +140,42 @@ export function rehydrateAsset(asset: IAssetInfo, moduleMap: IModuleMap, banner:
 
   source.add(suffix);
 
-  return new CachedSource(source);
+  const externals: Map<string, string> = new Map();
+  let nextOrdinal: number = 0;
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/RehydrateAsset.ts:144]: Same here.

File: webpack/module-minifier-plugin/src/ModuleMinifierPlugin.types.ts [hunk 4]
```diff
@@ -23,6 +23,10 @@ export interface IModuleMinificationRequest {
    * File name to show for the source code in the source map
    */
   nameForMap: string | undefined;
+  /**
+   * Reserved variable names, e.g. __WEBPACK_EXTERNAL_MODULE_1__
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/ModuleMinifierPlugin.types.ts:27]: External variable names don't necessarily have numerical identifiers. That's handled, right?

File: common/reviews/api/module-minifier-plugin.api.md (hunk 1 above)
Comment (dmichon-msft) [reply common/reviews/api/module-minifier-plugin.api.md:50]: This already exists. It's used by some internal webpack plugins.

File: webpack/module-minifier-plugin/src/RehydrateAsset.ts [hunk 5]
```diff
@@ -133,5 +140,42 @@ export function rehydrateAsset(asset: IAssetInfo, moduleMap: IModuleMap, banner:
 
   source.add(suffix);
 
-  return new CachedSource(source);
+  const externals: Map<string, string> = new Map();
```
Comment (dmichon-msft) [inline webpack/module-minifier-plugin/src/RehydrateAsset.ts:143]: I forgot to delete this. It's been replaced by the `externalNames` block

File: webpack/module-minifier-plugin/src/ModuleMinifierPlugin.types.ts (hunk 4 above)
Comment (dmichon-msft) [reply webpack/module-minifier-plugin/src/ModuleMinifierPlugin.types.ts:27]: Yeah, the computation of the id is exactly the code used to generate them in webpack.

**Pull request merged**